### DISCUSSION TITLE
Is it no longer possible to use `Literal` as a key type in a dictionary?

### DISCUSSION THREAD

**Discussion by tasuren** [microsoft/pyright#5478] [Q&A] [answered]
If I specify the `Literal` type of string as the key of the dictionary, there was no error before, but now I get an error saying that the override of `__init__` is not enough.
Why is this?

### Code
```python
from typing import Literal, TypeAlias

Mode: TypeAlias = Literal["easy", "normal", "hard"]

games = dict[Mode, list[int]](
  easy=[0, 1], normal=[2, 3], hard=[4, 5]
)

print(games)
```
### Error
```shell
$ pyright test.py
**/test.py:5:9 - error: No overloads for "__init__" match the provided arguments
    Argument types: (list[int], list[int], list[int]) (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

## Versions
pyright: v1.1.316

## Additional Context
If I use pyright 1.1.314 instead of 1.1.316, there is no error.

**Comment by cainmagi** [comment, marked as answer]
I have taken a look at the stubs of the STL.

```python
class dict(MutableMapping[_KT, _VT], Generic[_KT, _VT]):
    # __init__ should be kept roughly in line with `collections.UserDict.__init__`, which has similar semantics
    # Also multiprocessing.managers.SyncManager.dict()
    @overload
    def __init__(self) -> None: ...
    @overload
    def __init__(self: dict[str, _VT], **kwargs: _VT) -> None: ...
    @overload
    def __init__(self, __map: SupportsKeysAndGetItem[_KT, _VT]) -> None: ...
    @overload
    def __init__(self: dict[str, _VT], __map: SupportsKeysAndGetItem[str, _VT], **kwargs: _VT) -> None: ...
    @overload
    def __init__(self, __iterable: Iterable[tuple[_KT, _VT]]) -> None: ...
    @overload
    def __init__(self: dict[str, _VT], __iterable: Iterable[tuple[str, _VT]], **kwargs: _VT) -> None: ...
    # Next two overloads are for dict(string.split(sep) for string in iterable)
    # Cannot be Iterable[Sequence[_T]] or otherwise dict(["foo", "bar", "baz"]) is not an error
    @overload
    def __init__(self: dict[str, str], __iterable: Iterable[list[str]]) -> None: ...
    @overload
    def __init__(self: dict[bytes, bytes], __iterable: Iterable[list[bytes]]) -> None: ...
```

As shown above, when using initialization like `dict(kw1=.., kw2=...)`, this usage implies that `str` is the type of keywords. In other words, if the keyword type is defined before the initialization, it is reasonable to me that narrowing down the pre-defined `dict[Literal[...], _VT]` type to any overloads with `dict[str, _VT]` fails in your case.

In my opinion, the current behavior of `pyright` is consistent with the typing stubs file. If the usages shown by you should be supported, the one that should be changed is not `pyright` but the typing stubs of STL.

### Workaround

The following codes work without errors
```python
from typing import Literal, TypeAlias, reveal_type

Mode: TypeAlias = Literal["easy", "normal", "hard"]

games: dict[Mode, list[int]] = dict(
  easy=[0, 1], normal=[2, 3], hard=[4, 5]
)

reveal_type(games)  # dict[Literal["easy", "normal", "hard"], list[int]]
```
because the `dict` is initialized with an `str` keyword type. Then the keyword type gets narrowed down to `Mode`.

**Reply by tasuren** [reply]
I see. In the meantime, I will try to follow that answer.
Thank you for your answer.

**Comment by cainmagi** [comment]
Sorry, my workaround provided in the previous answer is problematic because I found the following issue just now:

```python
from typing import Dict, Literal
from typing_extensions import reveal_type

Mode = Literal["a", "b", "c"]

val: Dict[Mode, int] = dict(d=1, e=2, f=3)  # Suppose to raise errors, but actually not

reveal_type(val)  # dict[Literal["a", "b", "c"], int]
```

There should be an error raised in the assignment above. If the initialization of dict is changed to `{"d": 1, "e": 2, "f": 3}`, the error will be correctly raised.

I am not sure whether this is an issue. Apparently, the above assignment is allowed because `dict[str, int]` is narrowed down to `dict[Mode, int]`.
